cmd/squid-per-site-exporter/main.go

```diff
@@ -0,0 +1,283 @@
+	// Require TLS by default; explicitly disable to allow HTTP
+	tlsRequired = flag.Bool("web.tls-required",
+		getEnvDefault("WEB_TLS_REQUIRED", "true") == "true",
+		"Require TLS certificate and key to be present. If true and files are missing, the server will not start. (Env: WEB_TLS_REQUIRED)")
+
+	// Prometheus metrics
+	squidHitRatio = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "squid_site_hit_ratio",
+			Help: "Hit ratio per site (hits / (hits + misses))",
+		},
+		[]string{"hostname"},
+	)
+
+	squidHitTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "squid_site_hits_total",
+			Help: "Total number of cache hits per site",
+		},
+		[]string{"hostname"},
+	)
+
+	squidMissTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "squid_site_misses_total",
+			Help: "Total number of cache misses per site",
+		},
+		[]string{"hostname"},
+	)
+
+	squidRequestsTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "squid_site_requests_total",
+			Help: "Total number of requests per site",
+		},
+		[]string{"hostname"},
+	)
+
+	squidBytesTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "squid_site_bytes_total",
+			Help: "Total bytes transferred per site",
+		},
+		[]string{"hostname"},
+	)
+
+	squidResponseTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "squid_site_response_time_seconds",
+			Help:    "Response time per site in seconds",
+			Buckets: prometheus.DefBuckets,
+		},
+		[]string{"hostname"},
+	)
+)
+
+type SiteStats struct {
+	Hits     int64
+	Misses   int64
+	Bytes    int64
+	Requests int64
+}
+
+type Exporter struct {
+	mutex     sync.RWMutex
+	siteStats map[string]*SiteStats
+}
+
+func NewExporter() *Exporter {
+	return &Exporter{
+		siteStats: make(map[string]*SiteStats),
+	}
+}
+
+func (e *Exporter) parseLogLine(line string) {
+	// Squid log format: timestamp elapsedtime remotehost code/status bytes method URL rfc931 peerstatus/peerhost type
+	fields := strings.Fields(line)
+	if len(fields) < 7 {
+		return
+	}
+
+	// Extract relevant fields
+	elapsedTimeStr := fields[1]
+	codeStatus := fields[3]
+	bytesStr := fields[4]
+	method := fields[5]
+	urlStr := fields[6]
+
+	// Skip non-HTTP methods
+	if !strings.HasPrefix(method, "GET") && !strings.HasPrefix(method, "POST") &&
+		!strings.HasPrefix(method, "HEAD") && !strings.HasPrefix(method, "PUT") {
+		return
+	}
+
+	// Parse URL to extract hostname
+	parsedURL, err := url.Parse(urlStr)
+	if err != nil {
+		return
+	}
+
+	hostname := parsedURL.Hostname()
+	if hostname == "" {
+		return
+	}
+
+	// Parse bytes
+	bytes, err := strconv.ParseInt(bytesStr, 10, 64)
+	if err != nil {
+		bytes = 0
+	}
+
+	// Parse elapsed time
+	elapsedTime, err := strconv.ParseFloat(elapsedTimeStr, 64)
+	if err != nil {
+		elapsedTime = 0
+	}
+
+	// Determine hit/miss from status code
+	isHit := false
+	if strings.Contains(codeStatus, "HIT") || strings.Contains(codeStatus, "REFRESH") {
```

Comment:
From what I gather, the result codes in the log can be any combination of a number of well defined tags separated by underscores. See https://wiki.squid-cache.org/SquidFaq/SquidLogs#squid-result-codes.

I'm concerned that just checking for `REFRESH` anywhere in the code doesn't indicate a cache hit. For example, `TCP_REFRESH_MODIFIED` indicates squid reached out to the server to check for an updated version of the requested resource and found one.

What testing have you done to validate that this is correctly identifying cache hits?